lib/kvdb/indent.c

```diff
@@ -0,0 +1,213 @@
+{
+    *arg1 = 0;
+    **arg2 = 1;
+    return 0;
+}
+
+
+// Func Decls w/ Attributes
+
+void HSE_NORETURN
+fn_attr1(void);
+
+HSE_NORETURN void
+fn_attr2(void);
+
+void HSE_PRINTF(1, 2)
+fn_attr3(char *fmt);
+
+void HSE_NORETURN
+fn_attr4(unsigned int arg1, unsigned int arg2, unsigned int arg3, struct foobar_foobar *foobar);
+
+HSE_NORETURN void
+fn_attr5(unsigned int arg1, unsigned int arg2, unsigned int arg3, struct foobar_foobar *foobar);
+
+void HSE_PRINTF(1, 2)
+fn_attr6(unsigned int arg1, unsigned int arg2, unsigned int arg3, struct foobar_foobar *foobar);
+
+
+hse_err_t
+long_if_expr(int flags)
+{
+    // simple long expr
+    if (long_variable_boolean || long_variable_integer > 0 || long_variable_pointer != NULL ||
+        (flags & HSE_KVS_PUT_VCOMP_MASK) == HSE_KVS_PUT_VCOMP_MASK)
+        return merr(EINVAL);
+
+    // long expr w/ HSE_LIKELY, which looks like a function call to clang-format
+    if (HSE_UNLIKELY(
+            long_variable_boolean || long_variable_integer > 0 || long_variable_pointer != NULL ||
+            (flags & HSE_KVS_PUT_VCOMP_MASK) == HSE_KVS_PUT_VCOMP_MASK))
+        return merr(EINVAL);
+
+    // long expr w/ function call in the middle
+    if (long_variable_boolean ||
+        some_long_function_name(
+            long_variable_boolean || long_variable_integer > 0 || long_variable_pointer != NULL))
+        return merr(EINVAL);
+
+    return 0;
+}
+
+
+void
+long_fn_call(void)
+{
+    // short
+    longish_printf("%u %u %u\n", 1, 2, 3);
+
+    // medium
+    longish_printf(
+        "test123=%d test123=%d test123=%d\n", 1000 + 1000 + 1000, 1000 + 1000 + 1000, 1000 + 1000);
+
+    // long
+    longish_printf(
+        "test123=%d test123=%d test123=%d\n", long_variable_integer, long_variable_integer,
+        long_variable_integer);
+
+    // without eol comment
+    printf(
+        "kblk_alloc_ms=%u kblk_write_ops=%u kblk_write_sz=%u "
+        "kblk_write_ms=%u vblk_alloc_ops=%u vblk_alloc_sz=%u "
+        "vblk_alloc_ms=%u vblk_write_ops=%u vblk_write_sz=%u\n",
+        uintvec[0], uintvec[1], uintvec[2], uintvec[3], uintvec[4], uintvec[5], uintvec[6],
+        uintvec[7], uintvec[8]);
+
+    // with eol hack
+    printf(
+        "kblk_alloc_ms=%u kblk_write_ops=%u kblk_write_sz=%u "
+        "kblk_write_ms=%u vblk_alloc_ops=%u vblk_alloc_sz=%u "
+        "vblk_alloc_ms=%u vblk_write_ops=%u vblk_write_sz=%u\n",
+        uintvec[0], uintvec[1], uintvec[2], // eol
+        uintvec[3], uintvec[4], uintvec[5], // eol
+        uintvec[6], uintvec[7], uintvec[8]);
+}
+
+
+int
+labels2(void)
+{
+    extern int x1, x2, x3;
+
+    if (!x1)
+        goto invalid1;
+
+    if (!x2)
+        goto invalid2;
+
+    if (!x3)
+        goto invalid3;
+
+    return 0;
+
+invalid1:
+    return -1;
+
+invalid2: // eol
+    {
+    int i;
+    for (i = 0; i < HSE_MCLASS_COUNT; i++) {
+        printf("%d", i);
+    }
+}
+    return -2;
+
+invalid3 : {
+    int i;
+    for (i = 0; i < HSE_MCLASS_COUNT; i++) {
```

Comment:
Does clang-format have a knob for always using braces or not for these one line scopes? Should we standardize on something?

Reply:
Not sure what you mean by "one line scope".  How can you avoid using a brace for any scope?

Reply:
```c
for (int i = 0; i < 10; i++)
  do_stuff();

// vs

for (int i = 0; i < 10; i++) {
  do_stuff();
}
```

Reply:
I suspect he means elide the braces if there's only one statement in the block?  I would rather clang-format not be able to do that even if it could!!!

Reply:
AFAICT it does work off the AST, that's why it has lost all the user's formatting and can't get it back.  So it has a few rules on how to regurgitate it...

Reply:
Tristan wants clang-transmogify :-)

Reply:
See this comment in the docs for `InsertBraces` setting:
> Setting this option to true could lead to incorrect code formatting due to clang-format’s lack of complete semantic information. As such, extra care should be taken to review code changes made by this option.

Reply:
How lame, why wouldn't they use the clang ast???

clang -Xclang -ast-dump file.cpp